File: Python/linked_list.py
class Node:
    def __init__(self, value) -> None:
        self.value = value
        self.next: Node = None

class LinkedList:
    def __init__(self, head: Node = None) -> None:
        self.head = head

    def length(self) -> int:
        counter = 0
        cur = self.head

        if self.head:
            counter += 1
            
            while cur.next:
                cur = cur.next
                counter += 1
        return counter

    def append(self, new_element: Node) -> None:
        cur = self.head
        if self.head:
            while cur.next:
                cur = cur.next
            cur.next = new_element
        else:   
            self.head = new_element
    
    def value_at(self, index: int):
        cur = self.head
        idx = 0

        if cur:
            if index == idx:
                return cur.value

            while cur.next:
                cur = cur.next
                idx += 1
                
                if index == idx:
                    return cur.value
                
        return None      

    def node_at(self, index: int):
        cur = self.head
        idx = 0

        if cur:
            if index == idx:
                return cur

            while cur.next:
                cur = cur.next
                idx += 1
                
                if index == idx:
                    return cur
                
        return None     

    def insert_at(self, index: int, new_node: Node):
        cur = self.head
        idx = 0

        if index == 0:
            new_node.next = self.head
            self.head = new_node

        if self.head:
            while cur:
                cur = cur.next
                idx += 1
                
                if index == idx:
                    node_before_cur = self.node_at(idx - 1)
                    node_before_cur.next = new_node
                    new_node.next = cur

    
    def delete(self, value):
        cur = self.head
        idx = 0

        if value == cur.value:
            second_node = self.node_at(1)
            self.head = second_node
            cur.next = None

        while cur.next:
            cur = cur.next
            idx += 1
            
            if value == cur.value:
                prev_node = self.node_at(idx - 1)
                try:
                    next_node = self.node_at(idx + 1)
                    prev_node.next = next_node
                    cur.next = None
                except:
                    prev_node.next = None
    
    def update(self, index, new_value):
        node_to_update = self.node_at(index)
        
        try:
            node_to_update.value = new_value
        except:
            raise TypeError

l = LinkedList()

print("LENGTH OF LIST: ", l.length())

n1 = Node("FIRST NODE")
l.append(n1)

print("LENGTH OF LIST: ", l.length())


n2 = Node("SECOND NODE")
l.append(n2)

print("LENGTH OF LIST: ", l.length())


n3 = Node("THIRD NODE")
l.append(n3)

n4 = Node("FOURTH NODE")
l.append(n4)

n5 = Node("INSERT THIS")
l.insert_at(2, n5)

n6 = Node("I AM TOP")
l.insert_at(0, n6)

n7 = Node("I AM LAST")
l.insert_at(6, n7)

n8 = Node("NO, I AM THE LAST")
l.insert_at(7, n8)

print("LENGTH OF LIST: ", l.length())


"""
    0 - I AM TOP
    1 - FIRST NODE
    2 ...
    6 - FOURTH NODE
"""

print(l.value_at(0))
print(l.value_at(1))
print(l.value_at(2))
print(l.value_at(3))
print(l.value_at(4))
print(l.value_at(5))
print(l.value_at(6))
print(l.value_at(7))


l.delete("FIRST NODE")
l.delete("FOURTH NODE")

print(" <DELETED SOME> ")

print(l.value_at(0))
print(l.value_at(1))
print(l.value_at(2))
print(l.value_at(3))
print(l.value_at(4))
print(l.value_at(5))
print(l.value_at(6))
print(l.value_at(7))

print("LENGTH OF LIST: ", l.length())

print(" <UPDATED SOME> ")

l.update(0, "UPDATED_ TOP NODE")
l.update(1, "UPDATED BLA BLA")
l.update(l.length() - 1, "UPDATED LAST NODE")
#l.update(8, "ERROR")

print(l.value_at(0))
print(l.value_at(1))
print(l.value_at(2))
print(l.value_at(3))
print(l.value_at(4))
print(l.value_at(5))

File: Typescript/linkedList.ts
type NodeValue = string | number | boolean;

class NNode<NodeValue> {
    value: NodeValue;
    next?: NNode<NodeValue>;
    constructor(value: NodeValue) {
        this.value = value;
    }
}

class LinkedList {
    head?: NNode<NodeValue>;
    size: number = 0;

    nodeAt(index: number) {
        let cur = this.head;
        let curIndex = 0;

        if (cur) {
            if (index === curIndex) return cur;
            while (cur.next) {
                cur = cur.next;
                curIndex++;

                if (index === curIndex) return cur;
            }
        }
        return null;
    }

    prepend(node: NNode<NodeValue>) {
        if (this.size === 0) {
            this.head = node;
            return;
        } else node.next = this.head;
        this.head = node;
        this.size++;
    }

    append(node: NNode<NodeValue>) {
        if (this.size === 0) {
            this.head = node;
            this.size++;
        } else if (this.size === 1 && this.head !== undefined) {
            this.head.next = node;
            this.size++;
        } else {
            let cur = this.head;

            while (cur?.next) {
                cur = cur.next;
            }

            cur!.next = node;

            this.size++;
        }
    }

    insertAt(index: number, node: NNode<NodeValue>) {
        let cur = this.head;
        let curIndex = 0;

        if (index === 0) this.prepend(node);

        if (this.head) {
            while (cur) {
                cur = cur.next;
                curIndex++;

                if (index === curIndex) {
                    let prevNode = this.nodeAt(curIndex - 1);
                    prevNode!.next = node;
                    node.next = cur;
                    this.size++;
                }
            }
        }
    }

    pop(): NNode<NodeValue> | null {
        if (this.size > 1) {
            let toPop = this.nodeAt(0);

            if (toPop?.next) {
                this.head = toPop.next;
                toPop.next = undefined;
            } else {
                this.head = undefined;
            }

            return toPop;
        }
        return null;
    }

    delete(index: number) {
        if (index > this.size - 1) throw new Error("index out of range");
        if (this.size === 0) throw new Error("list is empty");
        if (index === 0) {
            this.pop();
            return;
        }

        let prevNode = this.nodeAt(index - 1);
        let nodeToDelete = this.nodeAt(index);

        if (nodeToDelete?.next) {
            prevNode!.next = nodeToDelete?.next;
            nodeToDelete.next = undefined;
        } else prevNode!.next = undefined;
    }
}

/***** TESTS *****/
const l = new LinkedList();

const n1 = new NNode<string>("FIRST NODE");
l.prepend(n1);
/*
console.log(l.size);
console.log(l.head!.value);
*/
const n2 = new NNode("BEFORE FIRST NODE");
l.prepend(n2);
/*
console.log(l.head!.value);
console.log(l.head!.next!.value);
*/

const n3 = new NNode("LAST");
l.append(n3);

//console.log(l.head?.next?.next?.value);
/*
console.log(l.nodeAt(0)?.value);
console.log(l.nodeAt(1)?.value);
console.log(l.nodeAt(2)?.value);
*/

const n4 = new NNode("INSERT AT");
//l.insertAt(0, n4)
//console.log(l.head?.value);

l.insertAt(2, n4);
//console.log(l.head!.next?.next?.value);
/*
l.delete(0);
console.log(l.head?.value);
console.log(l.head?.next?.value);
*/
/*
l.delete(1);
console.log(l.head?.value);
console.log(l.head?.next?.value);
*/
/*
l.delete(3);
console.log(l.head?.next?.next?.value);
console.log(l.head?.next?.next?.next);
*/
